### ISSUE TITLE
PAM: Should nullok removed from password also?

### ISSUE THREAD

**Issue by pyllyukko** [pyllyukko/harden.sh#64]
<https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/security_guide/chap-hardening_your_system_with_tools_and_services#sect-Security_Guide-Workstation_Security-Account_Locking>:

> To disable the `nullok` option, remove the `nullok` string from configuration files in the `/etc/pam.d/` directory, such as `/etc/pam.d/system-auth` or `/etc/pam.d/password-auth`.

<https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/managing_smart_cards/pam_configuration_files#Sample_PAM_Configuration_Files>:

>  The argument `nullok` instructs the module to allow the user to change their password *from* a blank password, otherwise a null password is treated as an account lock.

**Comment by pyllyukko** [comment]
<https://www.linuxquestions.org/questions/slackware-14/user-without-a-password-4175675708/#post6126043>

**Comment by pyllyukko** [comment]
`PAM_UNIX(8)` has an example with `password`:

```
password   required   pam_unix.so use_authtok nullok yescrypt
```

https://github.com/linux-pam/linux-pam/blob/cf2fc5ff7b4a8555fda2a5ebe5f6ab0e45c22996/modules/pam_unix/pam_unix_passwd.c#L562-L564 complains that "No password has been supplied."

**Comment by pyllyukko** [comment]
Somewhat related: 9d7cda3ad49aaa148e4a3abb6e51d4df53eaba14

**Comment by pyllyukko** [comment]
https://github.com/linux-pam/linux-pam/blob/ea980d991196df67cdd56b3f65d210b73218d08a/modules/pam_unix/pam_unix_passwd.c#L665-L669:

```C
/*
 * This is not an AUTH module!
 */
if (on(UNIX__NONULL, ctrl))
	set(UNIX__NULLOK, ctrl);
```

**Comment by pyllyukko** [comment]
[Disabling logins for user with null passwords](https://tldp.org/HOWTO/html_single/User-Authentication-HOWTO/#AEN295):

> You'll want to remove this argument from any modules of 'auth' type for services that allow login.

**Comment by pyllyukko** [comment]
Here's what Claude Opus 4.6 had to say:

# PAM `nullok`: Should It Be Removed from the `password` Stack?

**Date**: 2026-03-28
**Re**: [harden.yml#64](https://github.com/pyllyukko/harden.yml/issues/64)
**Source**: Linux-PAM @ commit `032fea2c` (`modules/pam_unix/`)

## TL;DR

**Removing `nullok` from the `password` stack has no effect.** The `pam_sm_chauthtok()` function unconditionally force-enables `nullok` behavior for password changes, regardless of whether `nullok` is specified in the PAM config. This has been the case since the initial commit (2000-06-20).

Removing `nullok` from the `auth` stack is what matters for security.

## The Code

In `pam_unix_passwd.c` lines 654–658 (unchanged since initial commit `ea488580` by Andrew G. Morgan, 2000-06-20):

```c
/*
 * This is not an AUTH module!
 */
if (on(UNIX__NONULL, ctrl))
    set(UNIX__NULLOK, ctrl);
```

This runs at the start of every `pam_sm_chauthtok()` call (password change), before any password verification.

## How the Flags Work

The `set()` macro is not a simple OR — it applies **both** a mask and a flag:

```c
#define set(x,ctrl)   ((ctrl) = ((ctrl)&unix_args[x].mask)|unix_args[x].flag)
```

The control table entries:

| Index | Token | Mask | Flag |
|-------|-------|------|------|
| `UNIX__NONULL` (9) | *(internal)* | `_ALL_ON_` | `01000` |
| `UNIX__NULLOK` (14) | `"nullok"` | `_ALL_ON_^(01000)` | `0` |

- **Default** (`UNIX_DEFAULTS`): `NONULL` is ON (bit `01000` set)
- **When `nullok` is parsed**: mask `_ALL_ON_^(01000)` clears the `NONULL` bit
- **`set(UNIX__NULLOK, ctrl)`**: applies the same mask, clearing `NONULL` — identical effect

## Proof: Both Paths Converge

Whether `nullok` is specified in the `password` line or not, `pam_sm_chauthtok()` produces the same result:

### Case 1: `nullok` NOT specified (default)

```
Initial:   NONULL=ON  (default)
Line 657:  on(UNIX__NONULL, ctrl) → true → enters if
Line 658:  set(UNIX__NULLOK, ctrl) → clears NONULL bit
Result:    NONULL=OFF  (nullok in effect)
```

### Case 2: `nullok` IS specified

```
Initial:   NONULL=OFF  (cleared by config parser)
Line 657:  on(UNIX__NONULL, ctrl) → false → skips
Result:    NONULL=OFF  (nullok in effect)
```

Both cases: **NONULL=OFF**, nullok behavior active. Verified with compiled [C test](https://github.com/user-attachments/files/26326455/pam_nullok_test.c) reproducing the exact flag logic.

## What This Means for Password Changes

After the override, `pam_sm_chauthtok()` calls `_unix_blankpasswd(pamh, ctrl, user)`:

1. With `NONULL` cleared, line 678 (`if (on(UNIX__NONULL, ctrl)) return 0`) does **not** short-circuit
2. If the user has a blank password → function returns 1 → `pam_sm_chauthtok` returns `PAM_SUCCESS` for the PRELIM phase (skips old password verification)
3. The user can then set a new password

If `NONULL` were still set (hypothetically), `_unix_blankpasswd` would return 0, and the user with a blank password would be asked for their old password, which would then fail verification — locking them out of changing it.

## `nullresetok` — The Auth-Stack Alternative

Added in commit `f5adefa1` (2019, Pavel Březina):

```c
// In _unix_blankpasswd(), called from pam_sm_authenticate():
if (on(UNIX_NULLRESETOK, ctrl)) {
    retval = _unix_verify_user(pamh, ctrl, name, &daysleft);
    if (retval == PAM_NEW_AUTHTOK_REQD) {
        pam_syslog(pamh, LOG_DEBUG,
            "user [%s] has expired blank password, enabling nullok", name);
        set(UNIX__NULLOK, ctrl);  // Same mechanism — clears NONULL via mask
    }
}
```

This uses the same `set(UNIX__NULLOK, ctrl)` mechanism to dynamically enable nullok during **authentication** — but only when the password is expired (`PAM_NEW_AUTHTOK_REQD`). This allows:

- Administrator creates account with blank expired password
- User authenticates (nullresetok allows it) and is immediately forced to change password
- After password is set, blank authentication is no longer possible

| Option | Stack | When nullok activates |
|--------|-------|-----------------------|
| `nullok` | auth | Always (any blank password) |
| `nullresetok` | auth | Only when password reset is enforced |
| *(none needed)* | password | Always (force-enabled by code) |

## Related Commits

| Commit | Description |
|--------|-------------|
| `ea488580` (2000) | Initial commit — `set(UNIX__NULLOK, ctrl)` in `pam_sm_chauthtok` |
| `f5adefa1` (2019) | Added `nullresetok` option |
| `8d0c5753` (2024) | Fixed empty password with non-empty hash edge case |
| `e6495e6a` (2024) | Enclosed macro arguments (cosmetic, behavior unchanged) |

## Recommendations for harden.yml

The current regex `\s+nullok(_secure)?` removes `nullok` from all PAM stack types (auth, account, password, session) in a file. This is fine:

1. **`auth` stack**: Removing `nullok` is the **important security measure** — prevents blank-password authentication
2. **`password` stack**: Removing `nullok` is **harmless but has no effect** — code overrides it
3. **`account` stack**: `nullok` has no role in account management
4. **`session` stack**: `nullok` has no role in session management

Consider using `nullresetok` on the `auth` stack if you need to support administrator-created accounts with blank expired passwords that must be changed on first login.

## Summary

| Question | Answer |
|----------|--------|
| Does `nullok` on `password` line affect behavior? | **No** — force-overridden by code |
| Is removing it from `password` harmful? | **No** — it's a no-op either way |
| Is removing it from `auth` important? | **Yes** — this is the security-relevant change |
| Can a user with blank password change it without `nullok`? | **Yes** — `pam_sm_chauthtok` always enables it |
| Is the PAM man page's example (`password required pam_unix.so nullok`) misleading? | **Yes** — the `nullok` there is redundant |
